### PR TITLE
[FEAT] Implement chain function. Resolves #79

### LINKED ISSUE #79: Implement chain function
Use the dynamic nature of **JavaScript Proxies** in Ruby by using **`method_missing`** meta-programming:

```rb
class ChainWrapper
  def initialize(value)
     @wrapped_value = value
  end

  def value
     @wrapped_value
  end

  def method_missing(method_name, *args, &block)
    "#{method_name} does not exist!"
  end
end

ChainWrapper.new([1,2,3]).islam('test')
```

### PR DESCRIPTION
**Pull request by Attrash-Islam**
This PR is related to implementing the `chain` function!

I Used the dynamic nature of **JavaScript Proxies** in Ruby by using _method_missing_ meta-programming and it went so damn good to be written in such an easy and clean way! 🌟 

_chain_ function Spec -> https://lodash.com/docs/4.17.15#chain

image

### CODE REVIEW

File: lib/utils/chain_wrpper.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            end
+
+            def value
+                @value
+            end
+
+            def method_missing(method_name, *args, &block)
+                begin
+                    method = @r_.method(method_name.to_s)
+                    result = method.(@value, *args)
```
Comment (Attrash-Islam) [inline lib/utils/chain_wrpper.rb]: The chained functions will not receive the data as the first argument (As Rudash functions usual expect) so we forward it from the wrapped value in the `ChainWrapper` class with the rest of the arguments! 🎉

**Comment by Attrash-Islam** [comment]
@thegedge  Do you have the time to review this tiny one? 🙂

File: lib/utils/chain_wrpper.rb (hunk 1 above)
Comment (thegedge) [inline lib/utils/chain_wrpper.rb]: I'd offer two suggestions here.

First, probably best to enforce visibility, so using `public_method` would be preferable over `method`.

Second, `public_send` is faster than `method` + `.()`, so I'd recommend collapsing these lines into:

```ruby
result = @r_.public_send(method_name, @value, *args)
```

And here's a benchmark I ran locally to show that it's faster:

```
Warming up --------------------------------------
       public_method   136.892k i/100ms
         public_send   220.832k i/100ms
Calculating -------------------------------------
       public_method      1.962M (± 1.8%) i/s -      9.856M in   5.025455s
         public_send      4.109M (± 0.8%) i/s -     20.758M in   5.052286s

Comparison:
         public_send:  4108924.9 i/s
       public_method:  1961890.5 i/s - 2.09x  slower
```

File: lib/utils/chain_wrpper.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+                    self.class.new(result, @r_)
+                rescue NameError => exception
```
Comment (thegedge) [inline lib/utils/chain_wrpper.rb]: nit: Since you're not using `exception`, you can just write it as

```ruby
rescue NameError
  ...
```

Unless this is your preferred style 🙂

File: lib/utils/chain_wrpper.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+                @value = value
+                @r_ = r_
+            end
+
+            def value
+                @value
+            end
+
+            def method_missing(method_name, *args, &block)
+                begin
```
Comment (thegedge) [inline lib/utils/chain_wrpper.rb]: FYI you can "attach" a `rescue` to `def` like this:

```ruby
def method
  ...
rescue NameError
  ...
end
```

Saves you a level of indentation, if you'd like that here.

File: test/chain.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+class ChainTest < Test::Unit::TestCase
+    def test_happy_path
+        filter_even = -> (number) { number % 2 == 0 }
+        inc_by_one = -> (number) { number + 1 }
+        result = R_.chain([1,2,3,4])
+                   .filter(filter_even)
+                   .map(inc_by_one)
+                   .head()
+                   .value()
+        assert_equal result, 3
```
Comment (thegedge) [inline test/chain.rb]: This is something I noticed in your tests. You have your assertions as

```ruby
assert_xyz actual, expected
```

whereas they should be written as

```ruby
assert_xyz expected, actual
```

This is because the failed message will have a message along the lines of `<A> expected but was <B>`. You may want to put up another PR to reverse all of these 🙂

File: lib/rudash.rb [hunk 5]
```diff
@@ -48,6 +48,9 @@
 require_relative './rudash/group_by.rb'
 require_relative './rudash/take.rb'
 require_relative './rudash/drop_right.rb'
+require_relative './rudash/chain.rb'
+
+require_relative './utils/chain_wrpper.rb'
```
[the remark this answers is not in the document]
Comment (Attrash-Islam) [reply lib/rudash.rb]: Will fix it!

File: lib/utils/chain_wrpper.rb (hunk 3 above)
Comment (Attrash-Islam) [reply lib/utils/chain_wrpper.rb]: Yeah, that would be perfect!

File: lib/utils/chain_wrpper.rb (hunk 1 above)
Comment (Attrash-Islam) [reply lib/utils/chain_wrpper.rb]: Will read more about these methods. Thanks!

File: test/chain.rb (hunk 4 above)
Comment (Attrash-Islam) [reply test/chain.rb]: Nice catch!
